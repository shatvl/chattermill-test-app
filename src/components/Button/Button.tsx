import styled from 'styled-components';

export enum ButtonSizes {
  XS = 'xs',
  SM = 'sm',
  MD = 'md',
  LG = 'lg',
}

const Button = styled.button<{
  size?: ButtonSizes | null;
  block?: boolean;
  active?: boolean;
  disabled?: boolean;
}>`
  display: inline-block;
  width: ${({ block }) => (block ? '100%' : 'initial')};
  vertical-align: middle;
  background: ${({ theme, active }) => (active ? theme.colors.primary : 'transparent')};
  border: 1px solid ${({ theme, active }) => (active ? theme.colors.primaryMedium : 'transparent')};
  font-size: ${({ size = ButtonSizes.MD }) => {
    switch (size) {
      case ButtonSizes.XS:
        return '0.875rem';
      case ButtonSizes.SM:
        return '0.875rem';
      case ButtonSizes.MD:
        return '1rem';
      case ButtonSizes.LG:
        return '1.125rem';
      default:
        return '0.875rem';
    }
  }};
  line-height: ${({ size = ButtonSizes.MD }) => {
    switch (size) {
      case ButtonSizes.XS:
        return '1rem';
      case ButtonSizes.SM:
        return '1.25rem';
      case ButtonSizes.MD:
        return '1.5rem';
      case ButtonSizes.LG:
        return '1.75rem';
      default:
        return '1.25rem';
    }
  }};
  padding: ${({ size = ButtonSizes.MD }) => {
    switch (size) {
      case ButtonSizes.XS:
        return '0.25rem 0.375rem';
      case ButtonSizes.SM:
        return '0.5rem 0.525rem';
      case ButtonSizes.MD:
        return '0.625rem 1.25rem';
      case ButtonSizes.LG:
        return '0.75rem 1.25rem';
      default:
        return '0.375rem 1rem';
    }
  }};
  border-radius: 0.25rem;
  color: ${({ theme, disabled }) => (disabled ? `${theme.colors.gray}` : theme.colors.primary)};
  box-sizing: border-box;
  outline: none;
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  transition: 0.15s all ease-in-out;
  text-decoration: none;
  white-space: nowrap;

  &:hover,
  &:active {
    outline: none;
  }

  &:hover {
    background-color: ${({ theme }) => theme.colors.gray200};
    border-color: ${({ theme }) => theme.colors.gray200};
  }

  &:active {
    background-color: ${({ theme }) => theme.colors.primaryMedium};
    border-color: ${({ theme }) => theme.colors.primaryLight};
  }

  &:disabled,
  &:disabled:hover,
  &:disabled:focus,
  &:disabled:active {
    cursor: not-allowed;
    color: ${({ theme }) => theme.colors.gray};
  }
`;

export default Button;
