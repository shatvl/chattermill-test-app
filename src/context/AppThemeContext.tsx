import React, { FC, createContext, useContext } from 'react';
import { DefaultTheme, ThemeProvider } from 'styled-components';

interface AppThemeContextProps {}

export interface AppTheme extends DefaultTheme {
  colors: {
    [key: string]: string;
  };
  zIndex: {
    [key: string]: number;
  };
  fonts: {
    header: string;
    main: string;
  };
  borderWidth: string;
}

// global theme for app, TODO: dark / light mode
export const getTheme = (): AppTheme => ({
  zIndex: {
    xs: 1,
    s: 2,
    m: 3,
    lg: 4,
    xlg: 5,
    xxlg: 6,
  },
  colors: {
    background: '#fff',
    primary: '#ff8489',
    primaryLight: 'rgba(255, 132, 136, .2)',
    primaryMedium: 'rgba(255, 132, 136, .5)',
    red: '#d65942',
    gray: '#c4c9cc',
    grayLight: 'e3e3e3',
    text: '#2e3939',
    textLighter: '#2e3939',
    textSecondary: '#707579',
    textMeta: '#bfbfbf',
    black: '#000000',
    white: '#ffffff',
  },
  fonts: {
    header: 'Hind Siliguri',
    main: 'Hind Siliguri',
  },
  borderWidth: '0.0625rem',
});

export const AppThemeContext = createContext<AppThemeContextProps>({});

export const AppThemeProvider: FC = ({ children }) => (
  //we could implement dark / light theme here, or user specific styles base on user session
  <AppThemeContext.Provider value={{}}>
    <ThemeProvider theme={getTheme()}>{children}</ThemeProvider>
  </AppThemeContext.Provider>
);

export const usePortalAppThemeContext = () => useContext(AppThemeContext);
