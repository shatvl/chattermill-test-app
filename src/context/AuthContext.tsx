import React, {
  FC, createContext, useContext, useState, useEffect,
} from 'react';
import jwt_decode from 'jwt-decode';
import { CHATTER_TOKEN_COOKIE, getCoockie } from '../services/cookie';
import { User } from '../models/User.model';

interface AuthContextProps {
  user: User;
  isAuthorized: () => boolean;
}

export const AuthContext = createContext<AuthContextProps>({
  user: null,
  isAuthorized: () => false,
});

export const AuthProvider: FC = ({ children }) => {
  const [user, setUser] = useState<User>();

  // decode token from params or try to retrieve from cookie
  const getUser = (token?: string): User => {
    const tok = token ?? getCoockie(CHATTER_TOKEN_COOKIE);

    if (!tok) {
      return null;
    }

    try {
      const decodedUser: { exp: number; id: string; orig_iat: number } = jwt_decode(tok);

      return decodedUser || null;
    } catch (error) {
      console.error(error);
      return null;
    }
  };

  const isAuthorized = () =>
    // check via getUser() method, because token could be expired and if users won't decoded navigate to login page
    !!getUser();
  useEffect(() => setUser(getUser()), []);

  return <AuthContext.Provider value={{ user, isAuthorized }}>{children}</AuthContext.Provider>;
};

export const useAuthContext = () => useContext(AuthContext);
