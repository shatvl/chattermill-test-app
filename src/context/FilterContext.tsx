import React, {
  FC, createContext, useContext, useState,
} from 'react';

interface FilterContextProps {
  themeId: number;
  setThemeId: (themeId: number) => void;
  clearFilters: () => void;
}

export const FilterContext = createContext<FilterContextProps>({
  themeId: null,
  setThemeId: () => {},
  clearFilters: () => {},
});

export const FilterProvider: FC = ({ children }) => {
  const [themeId, setThemeId] = useState<number>();

  const handleSelectThemeId = (themeId: number) => setThemeId(themeId);

  const handleClearFilters = () => setThemeId(null);

  return (
    <FilterContext.Provider value={{ themeId, setThemeId: handleSelectThemeId, clearFilters: handleClearFilters }}>
      {children}
    </FilterContext.Provider>
  );
};

export const useFilterContext = () => useContext(FilterContext);
