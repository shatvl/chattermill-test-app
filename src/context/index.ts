export { AppThemeContext } from './AppThemeContext';
export { usePortalAppThemeContext } from './AppThemeContext';
