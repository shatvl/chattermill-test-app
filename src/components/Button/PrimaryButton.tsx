import styled from 'styled-components';
import Button from './Button';

const PrimaryButton = styled(Button)`
  background: ${({ theme, active }) => (active ? theme.colors.primaryMedium : theme.colors.primary)};
  color: ${({ theme }) => theme.colors.white};
`;

export default PrimaryButton;
