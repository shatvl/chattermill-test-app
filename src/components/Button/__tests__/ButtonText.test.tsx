import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { AppThemeProvider } from '../../../context/AppThemeContext';
import Button, { ButtonSizes } from '../Button';
import { waitRenderer } from '../../../testing/wait';

describe('renders button', () => {
  it('default MD button check snapshot', async () => {
    const tree = renderer.create(
      <AppThemeProvider>
        <Button>click me</Button>
      </AppThemeProvider>,
    );

    await waitRenderer();

    expect(tree.toJSON()).toMatchSnapshot();
    expect(tree.toJSON()).toHaveStyleRule('font-size', '1rem');
    expect(tree.toJSON()).toHaveStyleRule('padding', '0.625rem 1.25rem');
  });

  it('extra small XS button check snapshot', async () => {
    const tree = renderer.create(
      <AppThemeProvider>
        <Button size={ButtonSizes.XS}>click me</Button>
      </AppThemeProvider>,
    );

    await waitRenderer();

    expect(tree.toJSON()).toMatchSnapshot();
    expect(tree.toJSON()).toHaveStyleRule('font-size', '0.875rem');
    expect(tree.toJSON()).toHaveStyleRule('padding', '0.25rem 0.375rem');
  });

  //TODO: continue for all cases :)
});
