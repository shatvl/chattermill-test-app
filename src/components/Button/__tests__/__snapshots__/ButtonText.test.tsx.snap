// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders button default MD button check snapshot 1`] = `
.c0 {
  display: inline-block;
  width: initial;
  vertical-align: middle;
  background: transparent;
  border: 1px solid transparent;
  font-size: 1rem;
  line-height: 1.5rem;
  padding: 0.625rem 1.25rem;
  border-radius: 0.25rem;
  color: #ff8489;
  box-sizing: border-box;
  outline: none;
  cursor: pointer;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c0:hover,
.c0:active {
  outline: none;
}

.c0:active {
  background-color: rgba(255,132,136,.5);
  border-color: rgba(255,132,136,.2);
}

.c0:disabled,
.c0:disabled:hover,
.c0:disabled:focus,
.c0:disabled:active {
  cursor: not-allowed;
  color: #c4c9cc;
}

<button
  className="c0"
>
  click me
</button>
`;

exports[`renders button extra small XS button check snapshot 1`] = `
.c0 {
  display: inline-block;
  width: initial;
  vertical-align: middle;
  background: transparent;
  border: 1px solid transparent;
  font-size: 0.875rem;
  line-height: 1rem;
  padding: 0.25rem 0.375rem;
  border-radius: 0.25rem;
  color: #ff8489;
  box-sizing: border-box;
  outline: none;
  cursor: pointer;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c0:hover,
.c0:active {
  outline: none;
}

.c0:active {
  background-color: rgba(255,132,136,.5);
  border-color: rgba(255,132,136,.2);
}

.c0:disabled,
.c0:disabled:hover,
.c0:disabled:focus,
.c0:disabled:active {
  cursor: not-allowed;
  color: #c4c9cc;
}

<button
  className="c0"
  size="xs"
>
  click me
</button>
`;
